File: ml/FareSystem/routes/fare_routes.py
"""
Fare Routes Blueprint
All fare calculation endpoints
"""
from flask import Blueprint, request, jsonify

# Blueprint will be initialized with fare_service in app.py
fare_bp = Blueprint('fare', __name__)

# Service will be set from app.py
fare_service = None


def init_fare_routes(service):
    """Initialize fare routes with service instance"""
    global fare_service
    fare_service = service


@fare_bp.route('/calculate_fare', methods=['POST'])
def calculate_fare():
    """
    Calculate bus fare based on boarding and alighting stages using fare matrix.
    Request body: { "boarding_stage": "Malabe", "alighting_stage": "Borella" }
    or { "boarding_stage_id": 3, "alighting_stage_id": 8 }
    """
    if not fare_service or not fare_service.fare_data:
        return jsonify({"error": "Fare data not loaded"}), 500
    
    data = request.get_json() or {}
    
    try:
        # Support both stage names and IDs
        boarding_id = data.get('boarding_stage_id')
        alighting_id = data.get('alighting_stage_id')
        
        # If names provided, convert to IDs
        if boarding_id is None and 'boarding_stage' in data:
            boarding_name = data['boarding_stage']
            boarding_id = next(
                (s['id'] for s in fare_service.fare_data['stages'] if s['name'] == boarding_name),
                None
            )
        
        if alighting_id is None and 'alighting_stage' in data:
            alighting_name = data['alighting_stage']
            alighting_id = next(
                (s['id'] for s in fare_service.fare_data['stages'] if s['name'] == alighting_name),
                None
            )
        
        # Validate inputs
        if boarding_id is None or alighting_id is None:
            return jsonify({"error": "Invalid boarding or alighting stage"}), 400
        
        result = fare_service.calculate_fare_by_stage(boarding_id, alighting_id)
        return jsonify(result), 200
        
    except ValueError as e:
        return jsonify({"error": str(e)}), 400
    except Exception as e:
        print(f"Fare Calculation Error: {e}")
        return jsonify({"error": str(e)}), 500


@fare_bp.route('/find_nearest_stage', methods=['POST'])
def find_nearest_stage():
    """
    Find the nearest bus stage to user's location.
    Request body: { "latitude": 6.9147, "longitude": 79.9729 }
    """
    if not fare_service or not fare_service.fare_data:
        return jsonify({"error": "Fare data not loaded"}), 500
    
    data = request.get_json() or {}
    user_lat = data.get('latitude')
    user_lon = data.get('longitude')
    
    if user_lat is None or user_lon is None:
        return jsonify({"error": "latitude and longitude are required"}), 400
    
    try:
        result = fare_service.find_nearest_stage_to_location(user_lat, user_lon)
        return jsonify(result), 200
        
    except ValueError as e:
        return jsonify({"error": str(e)}), 400
    except Exception as e:
        print(f"Find Nearest Stage Error: {e}")
        return jsonify({"error": str(e)}), 500


@fare_bp.route('/calculate_journey', methods=['POST'])
def calculate_journey():
    """
    Calculate complete journey including fare and time.
    Request body: {
        "boarding_stage_id": 3,
        "alighting_stage_id": 8,
        "bus_id": "NA-1234" (optional)
    }
    """
    if not fare_service or not fare_service.fare_data:
        return jsonify({"error": "Fare data not loaded"}), 500
    
    data = request.get_json() or {}
    
    try:
        boarding_id = data.get('boarding_stage_id')
        alighting_id = data.get('alighting_stage_id')
        bus_id = data.get('bus_id')
        
        if boarding_id is None or alighting_id is None:
            return jsonify({
                "error": "boarding_stage_id and alighting_stage_id are required"
            }), 400
        
        result = fare_service.calculate_journey(boarding_id, alighting_id, bus_id)
        return jsonify(result), 200
        
    except ValueError as e:
        return jsonify({"error": str(e)}), 400
    except Exception as e:
        print(f"Calculate Journey Error: {e}")
        return jsonify({"error": str(e)}), 500


@fare_bp.route('/calculate_fare_by_location', methods=['POST'])
def calculate_fare_by_location():
    """
    Calculate fare using GPS coordinates OR location names - automatically finds nearest stages.
    
    Request body OPTION 1 (Coordinates):
    {
        "boarding_location": {"latitude": 6.925, "longitude": 79.985},
        "alighting_location": {"latitude": 6.912, "longitude": 79.970},
        "bus_id": "NA-1234" (optional)
    }
    
    Request body OPTION 2 (Location Names):
    {
        "boarding_location": "SLIIT Malabe, Sri Lanka",
        "alighting_location": "Kaduwela Town, Sri Lanka",
        "bus_id": "NA-1234" (optional)
    }
    """
    if not fare_service or not fare_service.fare_data:
        return jsonify({"error": "Fare data not loaded"}), 500
    
    data = request.get_json() or {}
    
    try:
        boarding_loc_input = data.get('boarding_location')
        alighting_loc_input = data.get('alighting_location')
        bus_id = data.get('bus_id')
        
        if not boarding_loc_input or not alighting_loc_input:
            return jsonify({
                "error": "boarding_location and alighting_location are required"
            }), 400
        
        result = fare_service.calculate_fare_by_location(
            boarding_loc_input,
            alighting_loc_input,
            bus_id
        )
        return jsonify(result), 200
        
    except ValueError as e:
        return jsonify({"error": str(e)}), 400
    except Exception as e:
        print(f"Calculate Fare By Location Error: {e}")
        return jsonify({"error": str(e)}), 500

File: backend/routes/fareRoutes.js
const express = require('express');
const router = express.Router();
const axios = require('axios');

// Python ML Service URL (from environment or default)
const ML_SERVICE_URL = process.env.ML_SERVICE_URL || 'http://localhost:5000';

/**
 * POST /api/fare/calculate
 * Calculate fare between two bus stages
 * Body: { boarding_stage: "Malabe", alighting_stage: "Borella" }
 * or { boarding_stage_id: 3, alighting_stage_id: 8 }
 */
router.post('/calculate', async (req, res) => {
    try {
        const response = await axios.post(`${ML_SERVICE_URL}/calculate_fare`, req.body, {
            timeout: 5000
        });
        
        res.json(response.data);
    } catch (error) {
        console.error('Error calculating fare:', error.message);
        
        if (error.code === 'ECONNREFUSED') {
            return res.status(503).json({ 
                error: 'ML Service unavailable. Please ensure Python service is running on port 5000.',
                details: error.message 
            });
        }
        
        if (error.response) {
            return res.status(error.response.status).json(error.response.data);
        }
        
        res.status(500).json({ 
            error: 'Failed to calculate fare',
            details: error.message 
        });
    }
});

/**
 * GET /api/fare/route-info
 * Get complete route information including stages and fare matrix
 */
router.get('/route-info', async (req, res) => {
    try {
        const response = await axios.get(`${ML_SERVICE_URL}/get_route_info`, {
            timeout: 5000
        });
        
        res.json(response.data);
    } catch (error) {
        console.error('Error fetching route info:', error.message);
        
        if (error.code === 'ECONNREFUSED') {
            return res.status(503).json({ 
                error: 'ML Service unavailable. Please ensure Python service is running on port 5000.',
                details: error.message 
            });
        }
        
        if (error.response) {
            return res.status(error.response.status).json(error.response.data);
        }
        
        res.status(500).json({ 
            error: 'Failed to fetch route info',
            details: error.message 
        });
    }
});

/**
 * POST /api/fare/calculate-journey
 * Calculate complete journey with fare and bus status
 * Body: { boarding_stage_id: 3, alighting_stage_id: 8, bus_id: "NA-1234" }
 */
router.post('/calculate-journey', async (req, res) => {
    try {
        const response = await axios.post(`${ML_SERVICE_URL}/calculate_journey`, req.body, {
            timeout: 5000
        });
        
        res.json(response.data);
    } catch (error) {
        console.error('Error calculating journey:', error.message);
        
        if (error.code === 'ECONNREFUSED') {
            return res.status(503).json({ 
                error: 'ML Service unavailable. Please ensure Python service is running on port 5000.',
                details: error.message 
            });
        }
        
        if (error.response) {
            return res.status(error.response.status).json(error.response.data);
        }
        
        res.status(500).json({ 
            error: 'Failed to calculate journey',
            details: error.message 
        });
    }
});

/**
 * POST /api/fare/calculate-by-location
 * Calculate fare using GPS coordinates - perfect for locations BETWEEN stages!
 * Body: {
 *   boarding_location: { latitude: 6.925, longitude: 79.985 },
 *   alighting_location: { latitude: 6.912, longitude: 79.970 },
 *   bus_id: "NA-1234" (optional)
 * }
 */
router.post('/calculate-by-location', async (req, res) => {
    try {
        const response = await axios.post(`${ML_SERVICE_URL}/calculate_fare_by_location`, req.body, {
            timeout: 5000
        });
        
        res.json(response.data);
    } catch (error) {
        console.error('Error calculating fare by location:', error.message);
        
        if (error.code === 'ECONNREFUSED') {
            return res.status(503).json({ 
                error: 'ML Service unavailable. Please ensure Python service is running on port 5000.',
                details: error.message 
            });
        }
        
        if (error.response) {
            return res.status(error.response.status).json(error.response.data);
        }
        
        res.status(500).json({ 
            error: 'Failed to calculate fare by location',
            details: error.message 
        });
    }
});

/**
 * POST /api/fare/find-nearest-stage
 * Find the nearest bus stage to a GPS location
 * Body: { latitude: 6.925, longitude: 79.985 }
 */
router.post('/find-nearest-stage', async (req, res) => {
    try {
        const response = await axios.post(`${ML_SERVICE_URL}/find_nearest_stage`, req.body, {
            timeout: 5000
        });
        
        res.json(response.data);
    } catch (error) {
        console.error('Error finding nearest stage:', error.message);
        
        if (error.code === 'ECONNREFUSED') {
            return res.status(503).json({ 
                error: 'ML Service unavailable. Please ensure Python service is running on port 5000.',
                details: error.message 
            });
        }
        
        if (error.response) {
            return res.status(error.response.status).json(error.response.data);
        }
        
        res.status(500).json({ 
            error: 'Failed to find nearest stage',
            details: error.message 
        });
    }
});

module.exports = router;
